Streaming request body parsing

WyriHaximus opened the pull request:
This PR is the follow up for #13. It started out to make multipart streaming but ended up making all bodies streaming.

The parsers emit a `post` event with the key and value of a post variable and `file` on uploaded files found in the request. On the request object `getFiles` is gone due to the streaming nature of the parsers. `getPost` is still there but it won't have everything until the entire request has been parsed.

Todo:
- [x] Normal body streaming
- [x] Multipart body streaming
- [x] Form URL Encoded body streaming
- [x] File object for files in the request
- [x] Make sure all parsers behave the same
- [x] Make the form parsers optional
- [X] Add buffered sink like helpers for request post fields and files
- [X] Update readme with an example

nazar-pc commented:
I saw `gPsr`, are request/response objects PSR-7 compatible now?

WyriHaximus commented:
Not yet, those are also in the works though. But this is a step in that direction. The `gPsr` you're  referring to is only used to parse request headers.

clue commented:
> The parsers emit a post event with the key and value of a post variable and file on uploaded files found in the request.

The HTTP message body can potentially have any size (even single fields can be huge), do we really want to store this automatically?

As an alternative, nodejs only exposes the body stream and leaves it up to the consumer to pass this stream to the correct parser (for example http://stackoverflow.com/questions/4295782/how-do-you-extract-post-data-in-node-js).

WyriHaximus commented:
> The HTTP message body can potentially have any size (even single fields can be huge), do we really want to store this automatically?

No I want to go nodes way where a string or stream is emitted? Those can just be gathered using the buffered sink. Not sure yet about the stream part. It would result in less overhead for small fields but more code on the implementing side.

clue commented:
My personal vote here would be small, independent, composable parts instead of convenience built in.

Composable parts enable convenience on a higher level, such as this draft API:

```php
$http->on('request', function (Request $request, Response $response) use ($formParser) {
    $formParser->parseDeferredStream($request)->then(function ($fields) use ($response) {
       $response->end('hello ' . $field['name']);
   }, function ($error) use ($response) {
       $response->writeHead(400);
   });
});
```

Once we look into PSR-7 support, we could probably build a convenient middleware around this concept in order to make this available to each request handler.

WyriHaximus commented:
That looks good, I'm assuming that is just listening to `form` events emitted from `$request`. Or are you suggesting that `$request` just emits `data` and the form parser listens into that handling the parsing. Do like that even more :+1: . Should be fairly easy to do as well, I'll make sure both parsers behave the before refactoring into that

clue commented:
> suggesting that $request just emits data and the form parser listens into that handling the parsing. Do like that even more :+1:

This exactly :+1: 

> Should be fairly easy to do as well […]

Yeah, I too suppose this should be easier than auto-wiring all parsers :+1:

WyriHaximus commented:
Thanks for clarifying that @clue, working on that refactor right now :+1:

WyriHaximus commented:
@clue what I'm working out now would have this API approximately:

```php
$http->on('request', function (Request $request, Response $response) {
    FormParserFactory::create($request)->parseDeferredStream()->then(function ($fields) use ($response) {
       $response->end('hello ' . $field['name']);
   }, function ($error) use ($response) {
       $response->writeHead(400);
   });
});
```

Currently decoupling all that auto-wiring code

WyriHaximus commented:
The last few commits remove the right wiring between the form parsers and request parser. They also add a form parser factory. Next step is adding methods like `deferredStream` to them.

clue commented on src/FormParserFactory.php:
```diff
@@ -0,0 +1,25 @@
+        $headers = $request->getHeaders();
+
+        if (!array_key_exists('Content-Type', $headers)) {
+        }
```
?

WyriHaximus replied:
Left over/Use to return early once I figured out what to return when we can't detect a parser to use.

clue commented on src/FormParserFactory.php:
```diff
@@ -0,0 +1,25 @@
+{
+    public static function create(Request $request)
+    {
+        $headers = $request->getHeaders();
```
`http://php.net/array-change-key-case()` here to match case insensitive?

WyriHaximus replied:
Or #53, but `array_change_key_case` would work in there for now

WyriHaximus replied:
Done

clue commented on src/Parser/FormUrlencoded.php:
```diff
@@ -0,0 +1,56 @@
+        ) {
+            $this->buffer = substr($this->buffer, 0, $this->request->getHeaders()['Content-Length']);
+            $this->finish();
+        }
```
Is this needed here? If so, how about doing so once in the constructor?

marcj replied:
How about a `if (isset($this->request->getHeaders()['Content-Length']) && ...`? I think its more common than using `array_key_exists`, should also be faster.

WyriHaximus replied:
Fixed the `array_key_exists` usage and will move content size header detection to the constructor later.

clue commented on src/Parser/FormUrlencoded.php:
```diff
@@ -0,0 +1,56 @@
+        parse_str(trim($this->buffer), $result);
+        foreach ($result as $key => $value) {
+            $this->request->emit('post', [$key, $value]);
+        }
```
If its only responsibility is buffering the stream and then calling the parser once, how about using a `BufferedStreamSink` instead?

WyriHaximus replied:
Want to refactor this into streaming as well, haven't figured out the parsing for it yet

clue replied:
I think there's *some use* for parsing a multi-megabyte string here. However, the 80% use case is probably accessing these parsed fields from an array anyway. In this case, my vote would be to merely limit the stream size, then buffer its contents and invoke a simple function here instead.

Afaict this also appears similar to how PHP's SAPI handles this (`post_max_size` etc.).

A streaming parser can probably be implemented at a later stage (if still deemed useful).

WyriHaximus replied:
Fair enough, lets do this in a follow up PR :+1:

clue commented on src/Request.php:
```diff
@@ -30,6 +29,10 @@ public function __construct($method, $url, $query = array(), $httpVersion = '1.1
+
+        $this->on('post', function ($key, $value) {
+            $this->addPost($key, $value);
+        });
```
leftover?

WyriHaximus replied:
Correct

marcj commented on src/FormParserFactory.php:
```diff
@@ -0,0 +1,25 @@
+            return new Multipart($request);
+        }
+
+        if (strtolower($headers['Content-Type']) == 'application/x-www-form-urlencoded') {
```
Here we should check with strpos, because 'application/x-www-form-urlencoded' can have `; charset=utf8` as suffix.

clue commented on src/Parser/Multipart.php:
```diff
@@ -0,0 +1,190 @@
+ *
+ * @author jason.gerfen@gmail.com
+ * @author stephane.goetz@onigoetz.ch
+ * @license http://www.gnu.org/licenses/gpl.html GPL License 3
```
:-1: The GPL is not compatible with our MIT license.

(FWIW: Looks like this has been introduced as part of #13 originally)

marcj replied:
I really don't see anything in common anymore with the linked gist. I think its legit to just throw that `@license` away as those code files are meanwhile completely different.

WyriHaximus replied:
Does just removing `@license` gets us out of the woods or do I have to rewrite it to be sure? There are some things that also benefit from certain stream parts in that file.

WyriHaximus replied:
@cboden care to comment on this?

cboden replied:
re-write or use from #13. We can't mix licences and can't steal code that is licensed under GPL.

WyriHaximus replied:
A re-write it is, it was introduce via #13. My bad :disappointed:

pwhelan commented on src/FormParserFactory.php:
```diff
@@ -10,11 +10,12 @@ class FormParserFactory
 
-        $contentType = strtolower($headers['Content-Type']);
+        $contentType = strtolower($headers['content-type']);
 
```
Could something like this be applied in the RequestParser class for the Content-Length header? This would close an issue I had with ab (Apache Benchmark).

WyriHaximus replied:
There is no `Content-Length` in `RequestParser` anymore, it moved to `FormUrlencoded` where this was indeed applied: https://github.com/reactphp/http/blob/streaming-multipart/src/Parser/FormUrlencoded.php#L38-L39 :smile:

pwhelan replied:
good stuff! rush this PR now!

WyriHaximus replied:
Working as hard as time permits on it :smile:

marcj commented on src/Parser/Multipart.php:
```diff
@@ -0,0 +1,191 @@
+        $headers = array_change_key_case($headers, CASE_LOWER);
+        preg_match("/boundary=\"?(.*)\"?$/", $headers['content-type'], $matches);
+
+        $this->boundary = $matches[1];
```
Content-Type of `multipart/form-data` does not have a boundary. The body looks like

```
------WebKitFormBoundaryPr93i9I5JFCTMovV
Content-Disposition: form-data; name="file"; filename="Screen Shot 2016-01-14 at 21.53.53.png"
Content-Type: image/png

�PNG
```

This leads basically to a notice: `Notice: Undefined offset: 1`

WyriHaximus replied:
Good catch, I'll look into it :+1:

marcj commented:
Very very nice, can't wait to see this baby being merged :smile:

christoph-kluge commented on src/FormParserFactory.php:
```diff
@@ -0,0 +1,28 @@
+        $headers = $request->getHeaders();
+        $headers = array_change_key_case($headers, CASE_LOWER);
+
+        if (!array_key_exists('content-type', $headers)) {
```
empty condition over here?

WyriHaximus replied:
Haven't though out what to do what to return when we can't determine which parser to use. Not a fan of returning false and have mixed return types. Thinking of either returning a special class that does nothing or a way to check out front if we can return anything useful. /cc @clue

clue replied:
Given the current design, it probably makes sense to throw an `Exception` here, because it can not actually be "parsed as a form".

I'm not sure I particularly like the current design, however.

marcj replied:
`Content-Type` is optional in http spec, so we should really not throw an exception, because this it's not a exception. Better way is to treat it as if there is no body at all. However, this is already the case. All these comments (above and mine) refer to an old commit (d525755623ae8a2776ff196bae0d0523ebb77087)
See current: https://github.com/reactphp/http/pull/41/files#diff-5f9f68b3995267f69b2f776127df04f0

clue replied:
I'm well aware of the HTTP specs :) However this particular code does not deal with general purpose HTTP messages, but with "form messages" in particular. Note that I'm not suggesting throwing an Exception to the consumer of this lib, but merely within the above class. Given the current design, it makes sense to reject with a "not a form message".

See "vexing exceptions" for more details about why I'm not particularly happy with this design :)

WyriHaximus commented:
Been a while, and a bit longer then expected, but:
* Replaced the GPL multipart parser
* Emit multipart chunk headers with the `post` and `file` emits
* Fixed an issue brought up by @marcj in the comments

marcj commented:
Very nice! :)

christoph-kluge commented:
sweet!

marcj commented:
@WyriHaximus is this now ready to merge? :)

WyriHaximus commented:
@marcj Almost, going to write https://github.com/reactphp/http/pull/41#issuecomment-198762307 tomorrow evening or Teusday and then it's up to @clue doing it's git magic to get the GPL code out of the history.

marcj commented on src/Parser/NoBody.php:
```diff
@@ -0,0 +1,19 @@
+
+    public function __construct(Request $request)
+    {
+        Util::forwardEvents($this->request, $this, [
```
typo  -> `forwardEvents($request, ...`

WyriHaximus commented:
Just added the `DeferredStream` (not sure yet if it's the best name tbh):

```php
DeferredStream::create($parser)->then(function ($filesNfields) {
    $filesNfields = [
        'post' => [
            /* Multidimensional array matching the posted fields array */
            /* See https://github.com/reactphp/http/blob/streaming-multipart/tests/DeferredStreamTest.php#L64-L81 for an example of such array */ 
        ],
        'files' => [
            [
                'file' => File, // Instance of file containing it's name, type, and filename
                'file' => 'buffer contents', // String contents of the File in the same array
            ],
        ],
    ];
});
```

@marcj your comments on this would be much appreciated

marcj commented:
@WyriHaximus My code:

```php

        $parser = FormParserFactory::create($reactRequest);
        $deferred = new \React\Promise\Deferred();
        $promise = $deferred->promise();

        ppm_log('start parsing request');
        DeferredStream::create($parser)->then(function ($filesNfields) use ($reactRequest, &$post, &$files, $cookies, $headers, $deferred) {
            ppm_log('done parsing request');
            $class = $this->bootstrap->requestClass();
            
            $method = $reactRequest->getMethod();
            $query = $reactRequest->getQuery();

            /** @var SymfonyRequest $syRequest */
            $syRequest = new $class($query, $filesNfields['post'], $attributes = [], $cookies, $filesNfields['files'], $_SERVER);

            $syRequest->setMethod($method);
            $syRequest->headers->replace($headers);

            $deferred->resolve($syRequest);
        });

        return $promise;
```

I just fired a `curl http://127.0.0.1:8080`, but it never reaches `done parsing request`. Am I doing something wrong?

WyriHaximus commented:
Just check to be absolutely sure with a very simple example:

```php
<?php

use React\Http\DeferredStream;
use React\Http\FormParserFactory;

require 'vendor/autoload.php';

$loop = \React\EventLoop\Factory::create();
$socket = new React\Socket\Server($loop);

$http = new React\Http\Server($socket);
$http->on('request', function ($request, $response) {
    $parser = FormParserFactory::create($request);
    DeferredStream::create($parser)->then(function ($fieldsNfiles) {
        var_export($fieldsNfiles);
    });
    $response->writeHead(200, array('Content-Type' => 'text/plain'));
    $response->end("Hello World!\n");
});

$socket->listen(1337);
$loop->run();
```

And the following commands do yield a `var_export`ed array:
`curl http://localhost:1337/`
`curl -X post -d example=value http://localhost:1337/`
`curl -d "name=WyriHaximus&phone=123" http://localhost:1337/`

Currently also digging into file uploads which doesn't seem to be working with curl :angry:

WyriHaximus commented:
Yup found an interesting bug with curl and file uploads 0_o, digging into it and will report back with commits

WyriHaximus commented:
Small alteration of the example file, because I wrote and ended a response right after receiving it the uploaded file didn't came in. A.k.a. reactphp was faster writing a response then curl was sending the file using multipart:

```php
<?php

use React\Http\DeferredStream;
use React\Http\FormParserFactory;

require 'vendor/autoload.php';

$loop = \React\EventLoop\Factory::create();
$socket = new React\Socket\Server($loop);

$http = new React\Http\Server($socket);
$http->on('request', function ($request, $response) {
    $parser = FormParserFactory::create($request);
    DeferredStream::create($parser)->then(function ($fieldsNfiles) use ($response) {
        var_export($fieldsNfiles);
        $response->writeHead(200, array('Content-Type' => 'text/plain'));
        $response->end("Hello World!\n");
    });
});

$socket->listen(1337);
$loop->run();
```

But it isn't 100% working as it should yet so I'm debugging that right now

marcj commented on src/Parser/Multipart.php:
```diff
@@ -0,0 +1,251 @@
+        }
+        $this->request->on('data', [$this, $dataMethod]);
+        Util::forwardEvents($this->request, $this, [
+            'end',
```
I guess this is wrong. 'end' is called on request when the connection is closed, right? this explains maybe why the deferred parser only continues when the connection closes. Actually 'end' on the parser should only be called when the request body is sent, not when the connection closes.

marcj commented on src/Parser/NoBody.php:
```diff
@@ -0,0 +1,19 @@
+    public function __construct(Request $request)
+    {
+        Util::forwardEvents($request, $this, [
+            'end',
```
Same here. NoBody should rather just immediately trigger the 'end' event of ParserInterface instead of the request end, because the connection is then already closed and no need to answer the request :P

marcj commented on src/Parser/FormUrlencoded.php:
```diff
@@ -0,0 +1,71 @@
+        $this->request = $request;
+
+        $this->request->on('data', [$this, 'feed']);
+        $this->request->on('end', [$this, 'finish']);
```
Oh I guess here's the same like below

WyriHaximus commented:
Yeah that is part of the issue (and mistake on my end), the other part is that the closing boundary isn't detected properly

WyriHaximus commented:
Fixed the issue but somehow broke tests, so fixing those and also fixed the `NoBody` and `FormUrlencoded` parsers. The `FormUrlencoded` is interesting, because the `Content-Length` header always has to be send. So will also try and detect a new line.

WyriHaximus commented:
@marcj multi part fileupload works:

`curl --no-keepalive -F a=b -F c=d -F filedata=@example.php http://localhost:1337/`

Also made sure situations where we can't do anything with the body are properly catched by the `DeferredStream`

marcj commented on src/DeferredStream.php:
```diff
@@ -15,6 +15,13 @@ class DeferredStream
     public static function create(ParserInterface $parser)
     {
+        if ($parser->isDone()) {
+            return \React\Promise\resolve([
```
Actually I think throwing an exception would make more sense here. Failing silency is pretty nasty

WyriHaximus replied:
Actually that will be better, was trying to avoid exceptions a bit overzealous . Will refactor that shortly

marcj commented on src/FormParserFactory.php:
```diff
@@ -0,0 +1,37 @@
+            return new FormUrlencoded($request);
+        }
+
+        return new NoBody($request);
```
When I send `application/json` to my symfony application, symfony isn't able anymore to read the body because you dropped it when react/http doesn't know the content-type. Actually, a body should always be redirected to the application itself because it could be a custom encoding as well. I guess you can use NoBody only when `content-length: 0`

WyriHaximus replied:
Here is a quick hack up how to deal with JSON in the current situation:

```php
<?php

use React\Http\DeferredStream;
use React\Http\FormParserFactory;
use React\Http\Parser\NoBody;
use React\Promise\Deferred;

require 'vendor/autoload.php';

$loop = \React\EventLoop\Factory::create();
$socket = new React\Socket\Server($loop);

$http = new React\Http\Server($socket);
$http->on('request', function ($request, $response) {
    $parser = FormParserFactory::create($request);
    if ($parser instanceof NoBody) {
        $deferred = new Deferred();
        $promise = $deferred->promise();
        $buffer = '';
        $request->on('data', function ($data) use (&$buffer, $deferred, $request) {
            $buffer .= $data;
            $json = json_decode($buffer);
            if (json_last_error() == JSON_ERROR_NONE) {
                $request->removeAllListeners('data');
                $deferred->resolve($json);
            }
        });
    } else {
        $promise = DeferredStream::create($parser);
    }
    $promise->then(function ($result) use ($response) {
        var_export($result);
        $response->writeHead(200, array('Content-Type' => 'text/plain'));
        $response->end("Hello World!\n");
    });
});

$socket->listen(1337);
$loop->run();
```

`curl -H "Content-Type: application/json" -X POST -d '{"username":"xyz","password":"xyz"}' http://localhost:1337`

The main concern I have is that there is currently no way to detect the en of the body if content length is missing. Also in in the light of your other comment, dropping `NoBody` and throwing an exception instead could be an option. What is your take on doing that?

marcj replied:
> The main concern I have is that there is currently no way to detect the en of the body if content length is missing. 

That's ok. If no content-length is given, ignore the body (use NoBody). But if given we should read the body until end of content-length and add it for example to a `RawBody` instance. I guess `NoBody` should only be used if we can not detect how long the body might be (either no Content-Length or no Content-Type).

No exception please if we can't detect a body because this is still valid http. I guess all of your current classes like NoBody make perfectly sense, however, we need a addition class that addresses the need of a content-types reactphp/http is not aware of or when the client sends a body (content-length > 0) but we can not decode it.

With that I can check for those 2 instances (NoBody, RawBody or else) to know whether I need to use the parser

marcj replied:
Ok maybe even better to not have much boilerplate code in user land:

Add in DeferredStream a third return value `body` which is filled when we can not decode the body. If `NoBody` all three values ('post', 'files', 'body') is empty. This way a user of this library can easily handle majority of requests. This way I don't have to handle stuff like `if (parser instanceof NoBody)` but just get three values when I want to have a application that is able to parse incoming requests.

WyriHaximus replied:
Really like the `RawBody` idea, I'll work that out. But only through a flag on `FormParserFactory::create` ex. `FormParserFactory::create($request, true)` to ensure there is always the option to stream large body requests.

marcj replied:
Cool!

I don't understand your last point. I actually only want a `RawBody` if the incoming request has either

1. No Content-Type
2. A Content-Type reactphp/http can not automatically decode (means no `multipart/*', `application/x-www-form-urlencoded`)

`NoBody` should only be used when no Content-Length is given (because we can not detect the end of the request body).

I don't understand what a streamed request has to do with the RawBody. :) What do I get when second parameter is `false` and there's no content-type (or one that we do not decode)?

WyriHaximus replied:
I'll also research that `DeferredStream` option, it sounds good. Can be combined with `RawBody` very easily :+1:

WyriHaximus replied:
Hah sorry, guess our comments are a bit out of sync.

I'll implement `RawBody`, `FormParserFactory::create` will not have a second parameter and will consume `RawBody` adding the extra `body` index to the resulting result.

marcj replied:
hehe I guess it too. Awesome thanks! 😃

WyriHaximus commented:
Updated readme, and added `RawBody`

marcj commented on src/Parser/FormUrlencoded.php:
```diff
@@ -0,0 +1,66 @@
+        $headers = $this->request->getHeaders();
+        $headers = array_change_key_case($headers, CASE_LOWER);
+
+        $this->contentLength = $headers['content-length'];
```
needs a `isset` check

WyriHaximus replied:
And if not existing throw an exception I assume. That check is already done in the form parser factory.

marcj replied:
Oh sorry, I meant that same line in RawBody :)

Regarding FormUrlencoded: Actually I think when the Content-Type is form-urlencoded but content-length is not set, then it means there is nothing to parse. Which is I guess perfectly valid. The http server should throw as less as possible exceptions as I think there are very less real constraints of such decoding.

marcj commented:
Cool! I've made the change with the isset check to content-length manually and tested it. File uploading doesn't work for the following request (which is made in Chrome through XHR)

```
POST /jarves/admin/file/upload HTTP/1.1
Host: 127.0.0.1:8080
Connection: keep-alive
Content-Length: 132382
Origin: http://127.0.0.1:8080
X-Requested-With: XMLHttpRequest
X-FirePHP-Version: 0.0.6
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.94 Safari/537.36
Content-Type: multipart/form-data; boundary=----WebKitFormBoundaryJAOODe7UW3rzr0eW
Accept: */*
Referer: http://127.0.0.1:8080/jarves
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.8,de;q=0.6,ga;q=0.4
Cookie: PHPSESSID=foobar


------WebKitFormBoundaryJAOODe7UW3rzr0eW
Content-Disposition: form-data; name="file"; filename="Screen Shot 2016-05-09 at 22.21.22.png"
Content-Type: image/png


------WebKitFormBoundaryJAOODe7UW3rzr0eW
Content-Disposition: form-data; name="path"

/
------WebKitFormBoundaryJAOODe7UW3rzr0eW--
```

What happens is that 'post' is empty array, which means 'path' is not there. The file however has been read correctly.

marcj commented:
Just checked with a another file. With a ascii text file it works. However, it appears that the file content got a additional \n at the end. Content:

```
Hi

was geht?



ghihihih
```

With a newline at the end. In var_dump I see 
```
string(28) "Hi

was geht?



ghihihih

"
```

With uploading a text file instead of the image the post array is filled correctly. Weird :)

marcj commented on src/Parser/RawBody.php:
```diff
@@ -0,0 +1,33 @@
+        $headers = $request->getHeaders();
+        $headers = array_change_key_case($headers, CASE_LOWER);
+
+        $this->contentLength = $headers['content-length'];
```
This is what I meant. Content-Length is optional and it can be `0`, so we need to handle those cases :)

WyriHaximus replied:
So how do we handle those cases where `Content-Length` is `0`? Can add a `getBuffer` method but there is no guarantee the client is done sending contents

nazar-pc replied:
Honestly, I'd just skip body entirely when:
* `Content-Lenght` is `0`
* `Content-Length` is smaller than actual body

Every sane browser reports `Content-Length` correctly, so do sane programs. I found some specs quite flexible, but have doubts that 100% of the spec is absolutely necessary.

Also from the [spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html):
>  For compatibility with HTTP/1.0 applications, HTTP/1.1 requests containing a message-body MUST include a valid Content-Length header field unless the server is known to be HTTP/1.1 compliant. If a request contains a message-body and a Content-Length is not given, the server SHOULD respond with 400 (bad request) if it cannot determine the length of the message, or with 411 (length required) if it wishes to insist on receiving a valid Content-Length.

So 400/411 seems reasonable answer here.

marcj replied:
Take a look here: https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.13

> Any Content-Length greater than or equal to zero is a valid value. Section 4.4 describes how to determine the length of a message-body if a Content-Length is not given.

which refers to 4.4

> If a request contains a message-body and a Content-Length is not given, the server SHOULD respond with 400 (bad request) if it cannot determine the length of the message, or with 411 (length required) if it wishes to insist on receiving a valid Content-Length.

Pretty straightforward.  :) So instead of throwing a exception internal, we should send a response immediately.

marcj replied:
I'd prefer the later error, 411. We should also think about implementing Transfer-Encoding then 👍

WyriHaximus replied:
We should implement Transfer-Encoding, but that is something for a different PR :shipit:

marcj replied:
of course :)

WyriHaximus replied:
Right, I'm a tad stuck on this. The `411` is what we want, but how do we communicate that `411` back to the code without throwing an exception, or passing in the `Reponse` instance. One way could be to emit an error. Any ideas/suggestions/feedback @marcj?

andig replied:
You might leave it to FormParserFactory::create which could accept a response? It needs to understand the capabilities of the ParserInterface implementations anyway and could return NULL or false in this case?

marcj replied:
> The 411 is what we want, but how do we communicate that 411 back to the code without throwing an exception, or passing in the Reponse instance

We don't need to. reactphp/http is abstraction for the http protocol, which means it should be able to return own responses when a client request is made that is either not spec conform or when the http server doesn't understand fundamental protocol encoding like the current topic here. As a user of reactphp/http I have no benefit in knowing that a client made a bad request. My application should only get valid requests. To get a valid request reactphp/http should do whatever is necessary with the client based on http specs, so he sends us valid stuff. (If sending a 411/500/xxx response is part of this, then so be it)

clue replied:
Note that RFC 2616 is deprecated, see http://tools.ietf.org/html/rfc7230 and http://tools.ietf.org/html/rfc7231 instead.

> My application should only get valid requests.

I agree, this should be our main focus here :+1: 

Invalid requests should be rejected as per the spec, handling this as a consumer of this lib is probably out of scope.

My main issue with this ticket is that it tries to tackle multiple responsibilities. HTTP is a layered protocol and the newer specs make this much more obvious: Message framing (content-length, transfer-encoding etc.) should be handled *before* passing anything to the form parsers.

WyriHaximus replied:
> My main issue with this ticket is that it tries to tackle multiple responsibilities. HTTP is a layered protocol and the newer specs make this much more obvious: Message framing (content-length, transfer-encoding etc.) should be handled before passing anything to the form parsers.

How do you suggest we proceed? I'm going to let the `FormParserFactory` send out a `411` in case of invalid request. But how do we communicate that back to the user? A way could be using promises

clue replied:
Sending out a `411` in case the `Content-Length` header is *required* sounds sane. However, I'm currently failing to see *where* this actually *is* required in this ticket?

WyriHaximus replied:
It isn't and would be a nice bonus, perhaps finishing this PR and then do a followup adding that?

andig replied:
> It isn't and would be a nice bonus, perhaps finishing this PR and then do a followup adding that?

Strong +1. As php-pm user I control client and server both and am really interested in getting volkszaehler/httpd to install at all without messing with unstable dependencies. Anything that will get 0.5 out of the door would be highly appreciated.

WyriHaximus commented:
Interesting, I'll have a look at it. Does it work for normal chrome requests?

marcj commented:
Haven't tested it yet with "normal" `<form>` requests but I will find out tomorrow :)

WyriHaximus commented:
Confirmed the `\n` at the end of attached files with `<form>`. Using https://gist.github.com/WyriHaximus/37d632483793c6819b4c4c5c5581f1a6

andig commented on README.md:
```diff
@@ -31,3 +31,25 @@ This is an HTTP server which responds with `Hello World` to every request.
+
+## FormParserFactory and DeferredStream Usage
+
+The `FormParserFactory` parses a quest and determines which body parser to use (multipart, formurlencoded, raw body, or no body). Those body parsers emit events on `post` fields, `file` on files, and raw body emits `body` when it received the whole body. `DeferredStream` listens for those events and returns them through a promise when done.
```
Should be ...parses a request... Not a quest ;)

WyriHaximus replied:
Look at it from this angle, I'm on a quest adding streaming request bodies to `react/http` :stuck_out_tongue_closed_eyes: 

Thanks, will fix it :+1:

clue commented on src/Parser/ParserInterface.php:
```diff
@@ -0,0 +1,11 @@
+
+interface ParserInterface extends EventEmitterInterface
+{
+    public function __construct(Request $request);
```
The constructor must not be part of the interface. What is the intend here? Perhaps renaming this to a method also makes this more obvious?

WyriHaximus replied:
Just a regular `create` method or an static one?

clue replied:
The intent does not appear to be particularly clear, so what exactly is this interface supposed to do? (input parameters, expected output etc.)

WyriHaximus replied:
The intent is to parse a request and emit, `file`, `post`, `body` events.

marcj commented:
What else do we need to get this done? :) Better done than perfect.

WyriHaximus commented:
Running some browser tests to be sure Chrome/FF/IE are covered in both forms and file uploads. Aside from that it is up to @clue to do the rebase he had in mind to fix the GPL issue.

pwhelan commented:
Can this code handle chunked requests? If not I can get a PR up for that once it's merged.

WyriHaximus commented:
> Can this code handle chunked requests? If not I can get a PR up for that once it's merged.

No it doesn't, but your PR would be very welcome one this is merged :+1:

andig commented:
> Aside from that it is up to @clue to do the rebase he had in mind to fix the GPL issue.

Can we get going? php-pm and downstream applications are still depending on getting this done ;)

WyriHaximus commented:
Honestly I spent the last month prepping to speak at two conferences (currently riding home from one of them) plus unexpected family affairs requiring immediate attention. Don't have anything big coming in the coming months, so I will, after taking some rest tomorrow, get back to working on my big things OSS list. Where this PR is at the top of it ☺️.

picq commented:
@WyriHaximus While you are on this, which is awesome. Especially psr7, and handling post/files, that I had to do, any chance to include `json` parser ?

nazar-pc commented:
I'm strictly against at this point, better release as is sooner than later.

I already wrote stream-based multipart parser myself which is very fast and memory efficient (it is used for files uploading using XHR when also using custom HTTP methods like `UPLOAD). Now I just need stable version of React where I can get raw body bytes (preferable with automatic chunked transfers handling, but this is not critical) to plug it in.

In fact, having raw bytes it is trivial to add support for JSON or anything else.

picq commented:
@nazar-pc Sure, it is trivial, that is precisely why I don't consider that as blocker to release that we all are waiting for.
Its not up to me, but anyway stable release comes first

WyriHaximus commented:
@picq Thanks! We're not going to add JSON parsing on our end because that isn't the responsibility of the HTTP server. Same reason we use react/socket to handle the incoming connection. This package responsibility is to notify you of incoming requests, give you the tools to parse different kind of HTTP request formats, and provide a way to return a result to the requester. Dealing with content within those requests handed by this package (files, field fields, JSON strings etc) is the responsibility of the user.  

Just pushed a couple of commits to this PR ensuring proper multipart parsing (it had a few small flaws in it). Tested it on Chromium Linux, Firefox Linux and IE Windows, and they all work. 

As far as I'm concerned this PR is done, except for the fact it has merge issue with master I'll resolve tonight. Lets get this out the door :smile: .

andig commented:
There's one remaining conflict in composer.json due to the new dependencies:

    "react/promise": "^1.0|^2.0"
    "clue/block-react": "^1.1"

have been added. Those two can be manually resolved during the merge. Ready to go and release?

clue commented on composer.json:
```diff
@@ -8,7 +8,8 @@
         "react/stream": "^0.4",
-        "evenement/evenement": "^2.0"
+        "evenement/evenement": "^2.0",
+        "react/promise": "^1.0|^2.0"
```
Should be `^2.0 || ^1.1`

clue commented on src/File.php:
```diff
@@ -0,0 +1,74 @@
+    /**
+     * @return string
+     */
+    public function getName()
```
`getName()` vs `getFilename()` appears to be unclear?

nazar-pc replied:
This is how they are called in multipart headers, I'd say this is expected naming.

clue replied:
I'm not suggesting we *need* to change the names, merely that this is unclear and needs some documentation :+1: 

That being said, the "name" refers to the (HTML) form field name the file is uploaded as, while the "filename" refers to the actual name of the file. Is the "name" even part of the "file"?

nazar-pc replied:
This is how you get it from request body.
Imagine following:
```html
<input type="file" name="uploaded_file">
```

What you'll get in request body is:
```
Content-Disposition: form-data; name="uploaded_file"; filename="picture.jpg"
Content-Type: image/jpeg
```

And how you get it normally in PHP is `$_FILES['uploaded_file']`.

So yes, `name` is essential here. And I agree, that PhpDoc might have some sort of comment to clarify this, even though for me (since I have own implementation of multipart parser) this is very clear already.

clue replied:
Perhaps my previous remark was a bit unclear, "name" is certainly essential here because individual fields need to be identified :)

If you look at PHP's `$_FILES['uploaded_file']` structure, you'll see the following fields:

* `name` maps to our `getFilename()`
* `type` maps to our `getType()`
* `size` does not apply here because we're streaming the file contents
* `tmp_name` does not apply either
* `error` does not apply either

Noticeably, the "field name" is not part of the structure (which does make sense IMHO). This raises the question why we should add this as part of the `FileInterface`?

(Note I'm not saying there are no valid reasons for either decision, I merely want this to be deliberate and with proper documentation)

nazar-pc replied:
By the way, even though we are working with streams, at the time we've parsed one file and move to the next, we do actually know file size (honestly haven't read the whole implementation here why size is not included) and it will not be changed afterwards, so can be safely stored in object alongside with other info.

WyriHaximus replied:
Not always. When you upload small files that fit within a 4KB chunk you generally have a good change of parsing that file in one go. If you go beyond that the parser has no clue when that file will end and won't know until it reached the end. Now the deferred stream catches those and stores them in a buffer but the parser it self just emits a file with a readable stream for it's contents. So at the moment of finding the file it will emit the file and start processing it, without a clue how big it is at that time.

nazar-pc replied:
Got it, thanks for clarification!

WyriHaximus replied:
> If you look at PHP's `$_FILES['uploaded_file']` structure, you'll see the following fields:
> Noticeably, the "field name" is not part of the structure (which does make sense IMHO). This raises the question why we should add this as part of the `FileInterface`?

Another option would be to add it as argument to the `file` emit.

clue replied:
I think this makes sense :+1:

clue commented on src/File.php:
```diff
@@ -0,0 +1,74 @@
+    /**
+     * @return string
+     */
+    public function getType()
```
What "type" is this? :) Looks like this is the MIME-/content-type?

nazar-pc replied:
Again, in context of `File` class this is clearly MIME type of the file, like `image/jpeg` or `text/html`

clue replied:
Again, let's make this more obvious, either via documentation or a more descriptive name :+1: 

That being said, the header is called "Content-Type" :)

WyriHaximus replied:
Lets go for `getContentType` :+1:

clue commented on src/FormParserFactory.php:
```diff
@@ -0,0 +1,37 @@
+
+        if (
+            !isset($headers['content-type']) ||
+            (!isset($headers['content-type']) && !isset($headers['content-length']))
```
Is this condition intended?

WyriHaximus replied:
Yes but it could be simplified, it should have both a `type` and `length`.

clue replied:
Are you sure that's what it's testing? :)

How about `!isset($headers['content-type'], $headers['content-length'])` ?

WyriHaximus replied:
Actually that would fail 3 tests :X

WyriHaximus replied:
Yes I'm sure :)

clue commented on src/FormParserFactory.php:
```diff
@@ -0,0 +1,37 @@
+
+use React\Http\Parser;
+
+class FormParserFactory
```
Should this class be moved to the `Parser` namespace?
Also, perhaps consider renaming the namespace to be more descriptive?

WyriHaximus replied:
Might be a better place for it. Any suggestion on more descriptive names?

clue replied:
What does this "parse"? Not sure, how about a `FormParser` or `BodyParser` or `PayloadParser` namespace?

Then again, what does this class do?

WyriHaximus replied:
`BodyParser` sounds good to me, and then rename this class to `Factory` as it creates a parser for the determined content type.

clue commented on src/Parser/FormUrlencoded.php:
```diff
@@ -0,0 +1,66 @@
+        if (
+            $this->contentLength !== false &&
+            strlen($this->buffer) >= $this->contentLength
+        ) {
```
I feel like this is mixing concerns.. Shouldn't the whole body be delimited before passing it to the parsers?

clue commented on src/Parser/NoBody.php:
```diff
@@ -0,0 +1,15 @@
+use Evenement\EventEmitterTrait;
+use React\Http\Request;
+
+class NoBody implements ParserInterface
```
This concept looks a bit unclear to me, what is a `NoBody` parser supposed to do?

Looks like taking SRP one step too far… :)

clue commented on src/Parser/RawBody.php:
```diff
@@ -0,0 +1,33 @@
+    {
+        $this->buffer .= $data;
+
+        if (strlen($this->buffer) >= $this->contentLength) {
```
See also above for my concern about mixing responsibilities (DRY).

WyriHaximus commented:
@clue addressed your concerns about mixing responsibilities in 32edca3068dd09e3484654b7d40ac769039d290f

andig commented:
Checking back- all concerns addressed and ready to package?

jsor commented:
> Just added the DeferredStream (not sure yet if it's the best name tbh):

What about moving it to `StreamingBodyParser\BufferedSink` and model it like [streams BufferedSink](https://github.com/reactphp/stream/blob/master/src/BufferedSink.php)?

I'd also add a `Parser` or `StreamingParser` suffix the to parsers (StreamingParserInterface ->  MultipartStreamingParser) to be more clear and inline with the naming in react's components (LoopInterface -> StreamSelectLoop).

WyriHaximus commented:
Tbh I think `ParserInterface` is good, not to long but still clear on what it is combined with the namespace. `StreamingBodyParserInterface` and `StreamingBodyParsersuffix` would be to long IMHO.

jsor commented:
LGTM :shipit: 

Btw, absolutely awesome work on this PR ❤️

jsor commented on README.md:
```diff
@@ -31,3 +31,25 @@ This is an HTTP server which responds with `Hello World` to every request.
+
+## StreamingBodyParser\Factory and BufferedSink Usage
+
+The `FormParserFactory` parses a request and determines which body parser to use (multipart, formurlencoded, raw body, or no body). Those body parsers emit events on `post` fields, `file` on files, and raw body emits `body` when it received the whole body. `DeferredStream` listens for those events and returns them through a promise when done.
```
`FormParserFactory` -> `StreamingBodyParser\Factory`
`DeferredStream` -> `StreamingBodyParser\BufferedSink`

jsor commented on README.md:
````diff
@@ -31,3 +31,25 @@ This is an HTTP server which responds with `Hello World` to every request.
     $loop->run();
 ```
+
+## StreamingBodyParser\Factory and BufferedSink Usage
````
`BufferedSink` -> `StreamingBodyParser\BufferedSink`

jsor commented on README.md:
```diff
@@ -31,3 +31,25 @@ This is an HTTP server which responds with `Hello World` to every request.
+    $http = new React\Http\Server($socket);
+    $http->on('request', function ($request, $response) {
+        $parser = React\Http\StreamingBodyParser\Factory::create($request);
+        BufferedSink::createPromise($parser)->then(function ($result) use ($response) {
```
`BufferedSink` -> `React\Http\StreamingBodyParser\BufferedSink`

clue commented:
I've discussed this PR with @WyriHaximus and we both would love to get this in rather sooner than later :shipit: 

Unfortunately, this currently builds on top of some existing PRs which will be reverted via #59. This means that this PR will result in a *huge* merge conflict after that :-)

We'll look into getting #59 in first and will then file new PRs that supersede / replace this one. While we can't promise a timeframe, be assured this is currently one of our top priorities :+1:

WyriHaximus commented:
Yeah Github won't even let me merge it from the site :stuck_out_tongue_closed_eyes: . This PR is my top ReactPHP priority at the moment, would prefer to get https://github.com/reactphp/http-client/pull/58 in A.S.A.P. so I can fully focus on this right here.

I'll discus with @clue how exactly we're going to cut it up, but since most of the discussion has already taken place here we can move relatively quick

moe123 commented:
Hello gents,

if I may, some remarks about the adopted design :

-- Any request may have query parameters HEAD, PUT, POST, DELETE and GET... (1)
-- It will make sense to me to have an onEvent taking the HTTP verb as filter not "request"
-- You are inlining the buffer of uploaded files (Apply to BODY and PUT) (2)

- (1) _Thus, you are forcing the user to juggle between one to another object representation to get everything._
- (2) _which bars you to handle large file ; also a security threat. In my opinion it would be better to redirect the input to a FIFO file session then carefully extract and parse  per chunk from there ; everything in this safe space, even if the FS is involved  ; the full in memory solution is an utopia._

Suggestions :
_you should only pre-parse in memory the part-offsets from the FIFO file space ; then give the result totally parsed on demand ; files should be carefully extracted per chunk-unit-buffer (let's say 1024 or 2048) to their destination  if the user request it, not inlined in memory._

Regards.

WyriHaximus commented:
Been spending some time splitting this PR up into smaller ones, that resulted in the following pull requests:
#62: File object 
#69: Streaming body parsers foundation 
#70: Content length buffered sink 
#71: Parser: urlencoded (depends on #69 and #70)
#72: Parser: multipart (depends on #69 and #62)
#73: Streaming parser bufferedsink (depends on #69 and #62)

### -----------

#62: Uploaded File object that doesn't make sense on it's own but it provides something needed by #72 and #73.
#69: Provides the base factory and `NoBody` and `RawBody` parsers to get started.
#70: Mainly a buffered sink like the one from `react/stream` but cuts off at a set length or when the stream ends.
#71: Urlencoded parser that requires the base (#69) and the sink (#70).
#72: Multipart parser that requires the file object (#62) and parser factory (#69), will be updated once the factory is in and added to the factory.
#73: Buffered sink that takes a streaming parser and will resolve the promise when the parser is done parsing the incoming request.

### -----------

Order of merging (all PR's will be squashed on merge keeping the history clear): #69, #62, #70, #73, #71, #72 

### -----------

Will go over @moe123's comment carefully and see where adjustment is necessary. One of the things I've already done due to @moe123's is make all the parsers cancelable.

andig commented:
Checking back with how close we are to merging 😮. From what I understand PRs should be fine up to #62 and #69, open comment on #70 and open tasks on #71, #72, #73.

Anything userland could do to get closer to merging apart from friendly nagging?

WyriHaximus commented:
@andig yes #622 and #69 are done as far as I'm concerned, unless @jsor or @clue thinks otherwise.  And I like to get them in soon, I'll ping them on IRC tonight and see how they look at it.

Once that is in, start working on completing the other PR's. One of the issues I came across is that the urlencoded parser (#71) is going to be interesting as I can't use build in PHP functions to do the parsing without buffering.

andig commented:
> One of the issues I came across is that the urlencoded parser (#71) is going to be interesting as I can't use build in PHP functions to do the parsing without buffering.

Can't we assume- for the time being- that buffering for this case is ok, i.e. you either have a POST blob which doesn't need decoding or you have urlencoded data that will most likely not exceed a certain size?

WyriHaximus commented:
> > One of the issues I came across is that the urlencoded parser (#71) is going to be interesting as I can't use build in PHP functions to do the parsing without buffering.

> Can't we assume- for the time being- that buffering for this case is ok, i.e. you either have a POST blob which doesn't need decoding or you have urlencoded data that will most likely not exceed a certain size?

We could do that, I've set up several milestones that allows us to release this in parts. For example first getting the foundation out in `0.5.0`, then multipart in `0.5.1`, and then urlencoded parser in `0.5.2`.

bweston92 commented:
Any update?

WyriHaximus commented:
@bweston92 Check out this issue: https://github.com/reactphp/http/issues/120

kelunik commented on src/RequestParser.php:
```diff
@@ -28,27 +37,23 @@ public function feed($data)
             // Extract the header from the buffer
             // in case the content isn't complete
-            list($headers, $this->buffer) = explode("\r\n\r\n", $this->buffer, 2);
+            list($headers, $buffer) = explode("\r\n\r\n", $this->buffer, 2);
```
This might result in a large string operation. Better use the previous `strpos` and check that against the `$this->maxSize` before. You might also want to use `substr` as you already have the position then.
